refactor(shared-data): allow api level dependent flow rates

sfoster1 opened the pull request:
Recent hardware changes allow us to update the default flow rate for
Gen2 Single pipettes to 10mm/s plunger speed (resulting flow rate varies
per pipette class), including retroactively. To avoid changing user
protocols under them, this adds new data for the changed values:
- flow rate values now can be organized in an object keyed with api
levels
- this will eventually supercede the single-value structure
- the api level 2.0 key is the "default", and should be used unless
operating at a higher api level for which there is a superceding value.

## Review requests

- Does this structure look good?  Nothing  uses this data yet (that'll be a follow up pr)
- Do the values for api level 2.0 match the "value" defaults?
- Are the 2.6 values properly twice the 2.0 levels, where they exist?

sfoster1 committed:
refactor(shared-data): allow api level dependent flow rates

Recent hardware changes allow us to update the default flow rate for
Gen2 Single pipettes to 10mm/s plunger speed (resulting flow rate varies
per pipette class), including retroactively. To avoid changing user
protocols under them, this adds new data for the changed values:
- flow rate values now can be organized in an object keyed with api
levels
- this will eventually supercede the single-value structure
- the api level 2.0 key is the "default", and should be used unless
operating at a higher api level for which there is a superceding value.

sfoster1 commented on shared-data/pipette/definitions/pipetteNameSpecs.json:
```diff
@@ -8,22 +8,26 @@
       "homePosition": 220,
-      "travelDistance": 30
+      "travelDistance": 30,
+      "valuesByApiLevel": {"2.0": 5}
```
wrong place

sfoster1 committed:
fixup: keys inw rong p lace, formatting

IanLondon commented on shared-data/pipette/schemas/pipetteNameSpecsSchema.json:
```diff
@@ -12,6 +12,15 @@
       "type": "string",
       "enum": ["GEN1", "GEN2"]
+    },
+    "perApiLevelValues": {
```
`perApiLevelValues` should be `valuesByApiLevel`

this schema is permissive and doesn't have `"additionalProperties": false`, so it's not caught!

sfoster1 replied:
aha!

sfoster1 replied:
Wait this is the looked-up definition. `valuesByApiLevel` is the name of the objects.

IanLondon reviewed (changes requested):
JSON schema def not correct

sfoster1 committed:
fixup: python dev types needed i ntegratign

sfoster1 committed:
one last snapshot  update

sfoster1 committed:
fixup: fix the stupid schema

The entries for the flow rates were just not valid schemas in a weird
way that metaschema validation didn't catch. Cool. they're fixed now

sfoster1 committed:
fixup: probably shouldn't limit api  versions to one digit minors

Laura-Danielle reviewed (changes requested):
We either need to remove the ability to modify default flow rates in pipette settings (only available under DEV but customers still need to use it), or we need to find another way to categorize default speeds. If we are able to assert that pipette functions in GEN2 `2.0` hold up no matter what default speed they have, then maybe we should have _another_ schema definition that separates out default values by api level (for any parameter that may have a default which is not affected by pipette functions).

This may reduce the need for bumping model versions as well.

shlokamin commented on shared-data/pipette/schemas/pipetteNameSpecsSchema.json:
```diff
@@ -12,6 +12,29 @@
+    },
+    "configValue": {
+      "type": "object",
+      "required": ["value", "min", "max", "valuesByApiLevel"],
```
`valuesByApiLevel ` isn't required is it?

sfoster1 replied:
It is in this schema; the "2.0" entry contains the old default "value" level, so we can eventually remove the "value" key.

shlokamin commented on shared-data/python/opentrons_shared_data/pipette/dev_types.py:
```diff
@@ -38,6 +38,13 @@ class PipetteConfigElement(TypedDict):
+    value: float
+    min: float
+    max: float
+    valuesByApiLevel: Dict[str, float]
```
Idk how optional params work here but does it need to be specified that `valuesByApiLevel` is optional? If I recall correctly you can't explicitly say that one field is optional with a `TypedDict`, you have to make everything optional :/

sfoster1 replied:
It's not optional (see above)

sfoster1 commented:
> We either need to remove the ability to modify default flow rates in pipette settings 

I think this is the path forward. Flow rate modification is a first-class feature in the protocol API.

IanLondon commented on shared-data/pipette/schemas/pipetteNameSpecsSchema.json:
```diff
@@ -12,6 +12,29 @@
+      }
+    },
+    "configValue": {
+      "type": "object",
```
would you want to put `"additionalProperties": false,` inside `configValue`? Right now this is allowed:

```json
"defaultAspirateFlowRate": {
      "value": 5,
      "min": 0.001,
      "max": 50,
      "YOLOOOOO": 123,
      "valuesByApiLevel": { "2.0": 5 }
    },
```

sfoster1 replied:
I think we'd decided to avoid `additionalProperties: false` consistently to allow for future-compatible schema changes or something?

sfoster1 committed:
fixup: remove mutability for flow rate configs

IanLondon reviewed (approved):
looking good! I don't want to block on the additionalProperties thing but I'm not sure why we'd want to avoid it, since for this particular schema we can make "breaking changes" to the schema itself without version-desyncing trouble as long as they're backwards compatible (as this PR demonstrates)

Laura-Danielle reviewed (approved):
lgtm

sfoster1 committed:
fixup: handle flow rates not being mutable in pipette_config

While we don't have to load the new flow rate data structure yet, we
shouldn't choke on the old one

sfoster1 committed:
fixup: integration tests need to catch up
